devito/types/separable.py

```diff
@@ -0,0 +1,192 @@
+    >>> x, y, z = grid.dimensions
+    >>> SeparableFunction(name='a', grid=grid)
+    a_x(x)*a_y(y)*a_z(z)
+    >>> SeparableFunction(name='b', grid=grid, sep='sum')
+    b_x(x) + b_y(y) + b_z(z)
+    >>> SeparableFunction(name='c', grid=grid, sep='sum', separated=x)
+    c_x(x) + c_yz(y, z)
+    """
+    def __init__(self, *args, **kwargs):
+        # Is it sum or product separable
+        self._sep_op = spearation_op[kwargs.get('sep', 'prod')]
+
+        # Conventional Function inputs
+        self._grid = kwargs.get('grid')
+        self._name = kwargs.get('name')
+        self._space_order = kwargs.get('space_order', 1)
+
+        # Initialize subfunctions
+        # Separable dimensions
+        sep = self.separated_dimensions(**kwargs)
+        nonseparated = tuple(d for d in self.grid.dimensions if d not in sep)
+        dims = tuple(d for d in sep) + (nonseparated,)
+        self._dims = dims
+
+        func_list = [self.function(d, **kwargs) for d in sep]
+        if len(nonseparated) > 0:
+            func_list.append(self.function(nonseparated, **kwargs))
+
+        expr = self._sep_op(*func_list)
+
+        dims = tuple(d for d in sep) + (nonseparated,)
+        self._subfunctions = {d: f for d, f in zip(dims, func_list)}
+        self._dims = dims
+        self._expr = expr
+        self._fd = self.expr._fd
+
+    def function(self, dimensions, **kwargs):
+        dims = as_tuple(dimensions)
+        func_type = Function
+        names = "".join(d.name for d in dims)
+        shape = tuple(self.grid.dimension_map[d].loc for d in dims)
+        kwargs["name"] = '%s_%s' % (self.name, names)
+        func = func_type(dimensions=dims, shape=shape, **kwargs)
+        return func
+
+    @property
+    def sep_op(self):
+        return self._sep_op
+
+    @property
+    def grid(self):
+        return self._grid
+
+    @property
+    def space_order(self):
+        return self._space_order
+
+    def separated_dimensions(self, **kwargs):
+        return as_tuple(kwargs.get('separated', self.grid.dimensions))
+
+    @classmethod
+    def __dtype_setup__(cls, **kwargs):
+        return kwargs.get('grid').dtype
+
+    def __repr__(self):
+        return self.expr.__repr__()
+
+    __str__ = __repr__
+
+    def func(self, *args, **kwargs):
+        return self.from_expr(self.expr.func(*args, **kwargs))
+
+    @property
+    def evaluate(self):
+        return self.expr.evaluate
+
+    @property
+    def expr(self):
+        return self._expr
+
+    @property
+    def dims(self):
+        return self._dims
+
+    @property
+    def subfunctions(self):
+        return self._subfunctions
+
+    def __getitem__(self, dim):
+        return self.subfunctions[dim]
+
+    @property
+    def data(self):
+        return tuple(f.data for f in self.subfunctions.value)
+
+    def subs(self, *args, **kwargs):
+        new = super(SeparableFunction, self).subs(*args, **kwargs)
+        new._expr = self.expr.subs(*args, **kwargs)
```

Comment:
changing state, so this is wrong

I don't see why we need to overwrite subs or xreplace. If things are done properly, we should just be able to seamlessly leverage sympy's xreplace

maybe the issue is that ``expr`` isn't part of the ``args``  of the SeparableFunction ?

otherwise [it should just work](https://github.com/sympy/sympy/blob/master/sympy/core/basic.py#L1201)

EDIT -- I think I see what you're trying to work around. ``expr`` is constructed by ``__init__`` from some other arguments... so the solution is that either these other arguments are made part of ``args``, or (this is worse, though) we apply subs/xreplace to them, and then we return ``SeparableFunction(**new_kwargs)``